fix setup for sidekiq 6.1.0

hmaack opened the pull request:
https://github.com/mperham/sidekiq/compare/v6.0.7...v6.1.0 shows internal API changes. The setup needs to initialize the fetcher instead of passing the fetch class. 🤷‍♂️ 

https://github.com/mperham/sidekiq/issues/4614 

I tried to fix that the easy way.

hmaack committed:
fix setup for sidekiq 6.1.0

https://github.com/mperham/sidekiq/compare/v6.0.7...v6.1.0 shows internal API changes. The setup needs to initialize the fetcher instead of passing the fetch class. 🤷‍♂️

hmaack committed:
fixup! fix setup for sidekiq 6.1.0

hmaack commented:
fixed the rubocop issue but the CI has to run again 🕵️‍♂️

shlima commented:
Waiting for this a lot

jobwat commented:
is there any way to retrigger the CI? a empty commit?

schonert commented:
@hmaack awesome! 

Is there anything we can do to push this forward?

@jobwat it seems that it only runs once when the pull request is created. It would require a new PR.
https://github.com/sensortower/sidekiq-throttled/blob/master/.github/workflows/ci.yml#L7

hmaack committed:
make bulk_requeue callable when sidekiq is terminated

I found out after sidekiq terminates it will call the `bulk_requeue` method to ensure that every running job will get requeued which is called by the `CLI#launch` -> `Launcher#stop`. In the old sidekiq fetcher API ( < 6.1 ) it was necessary that the `bulk_requeue` is a class method. Now it has to be an instance method. Checkout this commit fce05c9d4b4c0411c982078a4cf3a63f20f739bc and the changes at the `BasicFetch` class.

hmaack commented:
The throttler gem supports versions below sidekiq 6.1, this PR is still not compatible for versions < 6.1. So in this case I would recommend either to check the version of sidekiq in the `#setup!` method and decide how to pass / setup the `Fetcher` or try another approach like to have a special `#setup` method for sidekiq >= 6.1. I would like to have an opinion from a contributer / maintainer about this. What do you think @ixti ? ✌️

gnomex commented:
Hey guys, I'm waiting for this merge, there is anything I can help you to merge as soon as possible?

currently using as 
```
git 'https://github.com/9feet/sidekiq-throttled', branch: 'fixup/fetcher-init-sidekiq-6.1.0' do
  gem "sidekiq-throttled"
end
```

ixti commented on lib/sidekiq/throttled/fetch.rb:
```diff
@@ -50,21 +50,19 @@ def retrieve_work
+      # @param [Array<Fetch::UnitOfWork>] units
+      # @return [void]
+      def bulk_requeue(units, _options)
+        return if units.empty?
```
This method must be on singleton class for sidekiq 5.x. Thus, probably it is worth to add a module that will be included in sidekiq 6.x and extended in sidekiq 5.x

hmaack replied:
Done without the conditional check but implemented as a module.

ixti commented on lib/sidekiq/throttled.rb:
```diff
@@ -62,7 +62,9 @@ def setup!
-          Sidekiq.options[:fetch] = Sidekiq::Throttled::Fetch
+          Sidekiq.options[:fetch] = Sidekiq::Throttled::Fetch.new(
+            Sidekiq.options
+          )
```
Can you please add a conditional here to check if we need one way or the other?

hmaack replied:
Done and tested 👍

ixti reviewed (changes requested):
Looks good to me, but I think we should do some conditionals for older and newer version of sidekiq

hmaack committed:
fixup! make bulk_requeue callable when sidekiq is terminated

hmaack committed:
add support of sidekiq 6.1

ixti commented on lib/sidekiq/throttled/fetch.rb:
```diff
@@ -12,6 +12,37 @@ module Throttled
+        end
+      end
+
+      include BulkRequeue
```
```suggestion
      module BulkRequeue
        # Requeues all given units as a single operation.
        #
        # @see http://www.rubydoc.info/github/redis/redis-rb/master/Redis#pipelined-instance_method
        # @param [Array<Fetch::UnitOfWork>] units
        # @return [void]
        def bulk_requeue(units, _options)
          return if units.empty?
          
          Sidekiq.logger.debug { "Re-queueing terminated jobs" }
          Sidekiq.redis { |conn| conn.pipelined { units.each(&:requeue) } }
          Sidekiq.logger.info("Pushed #{units.size} jobs back to Redis")
        rescue => e
          Sidekiq.logger.warn("Failed to requeue #{units.size} jobs: #{e}")
        end
      end
      
      # https://github.com/mperham/sidekiq/commit/fce05c9d4b4c0411c982078a4cf3a63f20f739bc
      if Gem::Version.new("6.1.0") <= Gem::Version.new(Sidekiq::VERSION)
        include BulkRequeue
      else
        extend BulkRequeue
      end
```

ixti commented on spec/sidekiq/throttled_spec.rb:
```diff
@@ -3,13 +3,22 @@
+        expect(Sidekiq.options[:fetch]).to be(Sidekiq::Throttled::Fetch)
+      else
+        expect(Sidekiq.options[:fetch]).to be_a(Sidekiq::Throttled::Fetch)
+      end
```
```suggestion
      if Gem::Version.new("6.1.0") <= Gem::Version.new(Sidekiq::VERSION)
        expect(Sidekiq.options[:fetch]).to be_a(Sidekiq::Throttled::Fetch)
      else
        expect(Sidekiq.options[:fetch]).to be Sidekiq::Throttled::Fetch
      end
```

ixti reviewed (changes requested):
LGTM, I'd like to simplify code a bit. And in specs, I'd better see `WorkingClassHero.perform_async(...)` with further queue check rather than `redis.rpush`. In order to avoid spec duplication in different versions I think we should add something like:

``` ruby
shared_examples "provides bulk requeue API" do |scope|
  ...
end

if Gem::Version.new("6.1.0") <= Gem::Version.new(Sidekiq::VERSION)
  describe "#bulk_requeue" do
    include_examples "provides bulk requeue API", fetcher
  end
else
  describe ".bulk_requeue" do
    include_examples "provides bulk requeue API", fetcher.class
  end
end

hmaack committed:
Update spec/sidekiq/throttled_spec.rb

simplify sidekiq compare in fetch spec

Co-authored-by: Alexey Zapparov <ixti@member.fsf.org>

hmaack committed:
Update lib/sidekiq/throttled/fetch.rb

at .bulk_requeue method as class or instance method dependening on sidekiq version

Co-authored-by: Alexey Zapparov <ixti@member.fsf.org>

hmaack committed:
update `bulk_requeue` implementation

- Fix the conditional check so that versions lower than 6.1 extend the `BulkRequeue` module and greater or equal 6.1 will include the `BulkRequeue` module
- Use shared_examples to provide the `bulk_requeue` API specs, unfortunately its not allowed to pass `let`, `subjects` to the shared example
- Use the `WorkingClassHero` dummy instead of Redis.push for the bulk_requeue setup spec

hmaack committed:
fixup! Update spec/sidekiq/throttled_spec.rb

hmaack commented:
Hi @ixti , thx for your suggestions. I switched all conditional checks to `Gem::Version.new(Sidekiq::VERSION) < Gem::Version.new("6.1.0")` because I got confused after reviewing the different styles in the code. I hope thats OK for you. Check bd2b2e0 and 0307564 they were also wrong at some points. So I tried to use a uniform approach.

ixti commented on spec/sidekiq/throttled/fetch_spec.rb:
```diff
@@ -50,6 +49,39 @@
 
+  shared_examples "provides bulk requeue API" do |scope|
+    before do
+      Sidekiq::Client.push_bulk({
+        "class" => WorkingClassHero,
+        "args"  => Array.new(3) { [1, 2, 3] }
+      })
```
I think this can be simplified to:

```suggestion
      3.times { |i| WorkingClassHero.perform_async(i) }
```

But I'm totally fine either way.

ixti commented on lib/sidekiq/throttled.rb:
```diff
@@ -93,6 +91,25 @@ def throttled?(message)
 
       private
 
+      # Preset the the sidekiq fetch strategy
+      #
+      # With sidekiq -> 6.1 the internal API of sidekiq changed slightly.
+      # Versions below sidekiq 6.1 needs to initialize the fetcher
+      # instead of passing the fetch class.
+      #
+      # @return [void]
+      def preset_strategy!
+        require "sidekiq/throttled/fetch"
+
+        Sidekiq.options[:fetch] = if Gem::Version.new(Sidekiq::VERSION) < Gem::Version.new("6.1.0")
+                                    Sidekiq::Throttled::Fetch
+                                  else
+                                    Sidekiq::Throttled::Fetch.new(
+                                      Sidekiq.options
+                                    )
+                                  end
+      end
```
Please rename the method to `#setup_strategy!`, use indented style for conditionals, and Reduce comment to just a link to the changeset that changed API :D

```suggestion
      # @return [void]
      def setup_strategy!
        require "sidekiq/throttled/fetch"

        # https://github.com/mperham/sidekiq/commit/fce05c9d4b4c0411c982078a4cf3a63f20f739bc
        Sidekiq.options[:fetch] =
          if Gem::Version.new(Sidekiq::VERSION) < Gem::Version.new("6.1.0")
            Sidekiq::Throttled::Fetch
          else
            Sidekiq::Throttled::Fetch.new(Sidekiq.options)
          end
      end
```

ixti reviewed (approved):
Thank you!

ixti commented:
Squash-merged in 700507c

ixti commented:
Released as [v0.13.0](https://github.com/sensortower/sidekiq-throttled/releases/tag/v0.13.0)
